cdap-app-fabric/src/main/java/co/cask/cdap/runtime/monitor/RuntimeMonitor.java

```diff
@@ -0,0 +1,162 @@
+
+  public RuntimeMonitor(String runtimeId, CConfiguration cConf, MessagePublisher messagePublisher,
+                        int pollTimeMs, ClientConfig clientConfig, int limit) {
+    this.runtimeId = runtimeId;
+    this.cConf = cConf;
+    this.messagePublisher = messagePublisher;
+    this.clientConfig = clientConfig;
+    this.restClient = new RESTClient(clientConfig);
+    this.limit = limit;
+    this.pollTimeMs = pollTimeMs;
+    this.topicsToMonitor = new HashMap<>();
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    LOG.debug("Starting runtime monitor {}....", runtimeId);
+    initializeTopics();
+  }
+
+  private void initializeTopics() throws Exception {
+    if (topicsToMonitor.isEmpty()) {
+      HttpResponse response = restClient.execute(HttpMethod.GET, clientConfig.getConnectionConfig().resolveURI
+                                                   (clientConfig.getApiVersion() +
+                                                      "/runtime/monitor/topics").toURL(),
+                                                 clientConfig.getAccessToken(),
+                                                 HttpURLConnection.HTTP_NOT_FOUND);
+      if (response.getResponseCode() == HttpURLConnection.HTTP_NOT_FOUND) {
+        throw new NotFoundException(response.getResponseBodyAsString());
+      }
+
+      String topicsToMonitor = response.getResponseBodyAsString(Charset.forName("UTF-8"));
+      this.topicsToMonitor = GSON.fromJson(topicsToMonitor, MAP_STRING_STRING_TYPE);
+    }
+
+    // TODO initialize from offset table for a given runtimeId
+    for (String topic : topicsToMonitor.values()) {
+      topicToRequest.put(topic, new MonitorConsumeRequest(topic, null, limit));
+    }
+  }
+
+  @Override
+  protected void run() throws Exception {
+    runThread = Thread.currentThread();
+
+    try {
+      while (!stopped) {
+        HttpResponse response = restClient.execute(HttpMethod.GET, clientConfig.getConnectionConfig().resolveURI
+                                                     (clientConfig.getApiVersion() + "/runtime/metadata").toURL(),
+                                                   GSON.toJson(topicToRequest), ImmutableMap.<String, String>of(),
+                                                   clientConfig.getAccessToken(), HttpURLConnection.HTTP_NOT_FOUND);
+
+        if (response.getResponseCode() == HttpURLConnection.HTTP_NOT_FOUND) {
+          throw new NotFoundException(response.getResponseBodyAsString());
+        }
+
+        Map<String, List<MonitorMessage>> monitorResponses =
+          GSON.fromJson(response.getResponseBodyAsString(Charset.forName("UTF-8")), MAP_STRING_MESSAGE_TYPE);
+
+        processResponse(monitorResponses);
+        Thread.sleep(pollTimeMs);
+      }
+    } catch (InterruptedException e) {
+      // Interruption means stopping the service.
+    } catch (Exception e) {
+      LOG.error("Error while fetching monitoring data from runtime {} ", runtimeId, e);
```

Comment:
The runtime monitor should never terminate regardless of what error it is, otherwise no one will be monitoring the runtime.